File: backend/app/agents/journeys.py
import json
from pathlib import Path
from datetime import datetime, timezone
from typing import List, Dict, Any, Optional
from ..config import DATA_DIR
from ..db import get_db
from .journal import mood_history

CATALOG_PATH = DATA_DIR / "journeys.json"

def _now_iso() -> str:
    return datetime.now(timezone.utc).isoformat()

def _load_catalog() -> Dict[str, Any]:
    if not CATALOG_PATH.exists():
        return {"journeys": [], "days": []}
    try:
        with open(CATALOG_PATH, "r", encoding="utf-8") as f:
            return json.load(f)
    except Exception:
        return {"journeys": [], "days": []}

def list_journeys() -> List[Dict[str, Any]]:
    catalog = _load_catalog()
    return catalog.get("journeys", [])

def get_journey(journey_id: str) -> Optional[Dict[str, Any]]:
    journeys = list_journeys()
    return next((j for j in journeys if j.get("id") == journey_id), None)

def get_journey_day(journey_id: str, day_number: int) -> Optional[Dict[str, Any]]:
    catalog = _load_catalog()
    days = catalog.get("days", [])
    return next((d for d in days if d.get("journeyId") == journey_id and d.get("dayNumber") == day_number), None)

def list_user_journeys(user_id: str) -> List[Dict[str, Any]]:
    db = get_db().read()
    progress_map = db.get("journeyProgress", {})
    return progress_map.get(user_id, [])

def start_journey(user_id: str, journey_id: str) -> Dict[str, Any]:
    journey = get_journey(journey_id)
    if not journey:
        raise ValueError("Journey not found")
        
    db = get_db().read()
    progress_map = db.setdefault("journeyProgress", {})
    user_list = progress_map.setdefault(user_id, [])
    
    existing = next((p for p in user_list if p.get("journeyId") == journey_id), None)
    if existing:
        # Backfill grace fields if missing
        if "graceDaysRemaining" not in existing:
            existing["graceDaysTotal"] = 2
            existing["graceDaysRemaining"] = 2
            existing["graceDaysUsed"] = 0
            existing["isGraceProtected"] = True
            existing["graceHistory"] = []
            existing["status"] = "active"
            get_db().write(db)
        return existing
        
    prog = {
        "journeyId": journey_id,
        "currentDay": 1,
        "completedDays": [],
        "startedAt": _now_iso(),
        "graceDaysTotal": 2,
        "graceDaysRemaining": 2,
        "graceDaysUsed": 0,
        "isGraceProtected": True,
        "graceHistory": [],
        "status": "active"
    }
    user_list.append(prog)
    get_db().write(db)
    return prog

def use_grace_day(user_id: str, journey_id: str, reason: str = "Intentional Rest / Sabbath") -> Dict[str, Any]:
    """
    Activates a Journey Grace Day.
    Protects the user's progress on currentDay and ensures no streak reset or failure occurs.
    """
    db = get_db().read()
    progress_map = db.get("journeyProgress", {})
    user_list = progress_map.get(user_id, [])
    prog = next((p for p in user_list if p.get("journeyId") == journey_id), None)
    if not prog:
        # If not started yet, auto-start
        prog = start_journey(user_id, journey_id)
        # re-read
        db = get_db().read()
        user_list = db.get("journeyProgress", {}).get(user_id, [])
        prog = next((p for p in user_list if p.get("journeyId") == journey_id), prog)

    rem = prog.get("graceDaysRemaining", 2)
    used = prog.get("graceDaysUsed", 0)

    if rem > 0:
        prog["graceDaysRemaining"] = rem - 1
        prog["graceDaysUsed"] = used + 1
    else:
        # Compassionate grace extension
        prog["graceDaysUsed"] = used + 1

    history = prog.setdefault("graceHistory", [])
    entry = {
        "id": f"grace-{len(history) + 1}",
        "timestamp": _now_iso(),
        "reason": reason,
        "protectedDay": prog.get("currentDay", 1),
        "scripture": "The Lord's mercies are new every morning; great is your faithfulness. (Lamentations 3:22-23)",
        "streakProtected": True
    }
    history.append(entry)
    prog["status"] = "grace_paused"
    prog["isGraceProtected"] = True

    get_db().write(db)
    return {
        "progress": prog,
        "graceEntry": entry,
        "message": f"Grace Day applied with zero penalty. Day {prog.get('currentDay', 1)} is safely held."
    }

def complete_grace_catchup(user_id: str, journey_id: str) -> Dict[str, Any]:
    """
    Resumes the journey seamlessly after a Grace Day.
    Offers a pressure-free return to the current day without backlog overwhelm.
    """
    db = get_db().read()
    progress_map = db.get("journeyProgress", {})
    user_list = progress_map.get(user_id, [])
    prog = next((p for p in user_list if p.get("journeyId") == journey_id), None)
    if not prog:
        raise ValueError("Journey not found for this user")

    prog["status"] = "active"
    get_db().write(db)
    return {
        "progress": prog,
        "message": f"Welcome back into God's presence! You are resuming Day {prog.get('currentDay', 1)} in peace."
    }

def simulate_missed_day(user_id: str, journey_id: str) -> Dict[str, Any]:
    """
    Simulation utility for testing: simulates missing a devotional day.
    Instead of penalty, a Journey Grace Day steps in automatically to preserve momentum.
    """
    return use_grace_day(user_id, journey_id, reason="Simulated Life Interruption / Auto-Protection")

def get_active_journey(user_id: str) -> Optional[Dict[str, Any]]:
    in_progress = [p for p in list_user_journeys(user_id) if not p.get("completedAt")]
    if not in_progress:
        return None
    prog = in_progress[-1]
    # Backfill default grace fields if needed
    if "graceDaysRemaining" not in prog:
        prog["graceDaysTotal"] = 2
        prog["graceDaysRemaining"] = 2
        prog["graceDaysUsed"] = 0
        prog["isGraceProtected"] = True
        prog["graceHistory"] = []
        prog["status"] = "active"

    journey = get_journey(prog.get("journeyId", ""))
    if not journey:
        return None
    day = get_journey_day(prog.get("journeyId", ""), prog.get("currentDay", 1))
    return {
        "journey": journey,
        "progress": prog,
        "day": day,
        "graceProtection": {
            "total": prog.get("graceDaysTotal", 2),
            "remaining": prog.get("graceDaysRemaining", 2),
            "used": prog.get("graceDaysUsed", 0),
            "status": prog.get("status", "active"),
            "isProtected": True,
            "recentHistory": prog.get("graceHistory", [])[-3:]
        }
    }

def complete_day(user_id: str, journey_id: str) -> Dict[str, Any]:
    db = get_db().read()
    progress_map = db.get("journeyProgress", {})
    user_list = progress_map.get(user_id, [])
    prog = next((p for p in user_list if p.get("journeyId") == journey_id), None)
    if not prog:
        raise ValueError("Journey not started for this user")
        
    journey = get_journey(journey_id)
    if not journey:
        raise ValueError("Journey not found")
        
    curr = prog.get("currentDay", 1)
    completed = prog.setdefault("completedDays", [])
    if curr not in completed:
        completed.append(curr)
        
    if curr >= journey.get("totalDays", 5):
        prog["completedAt"] = _now_iso()
    else:
        prog["currentDay"] = curr + 1
        
    get_db().write(db)
    return prog

def get_recommended_journey(user_id: str) -> Dict[str, Any]:
    journeys = list_journeys()
    if not journeys:
        return {
            "journey": {
                "id": "peace-in-anxiety",
                "title": "Peace in the Storm",
                "description": "5-day devotional anchor for anxiety and work stress.",
                "category": "Peace",
                "totalDays": 5,
                "recommendedMoods": ["peaceful", "seeking"]
            },
            "reason": "Popular foundation track",
            "personalized": False
        }
        
    user_progress = list_user_journeys(user_id)
    started_ids = {p.get("journeyId") for p in user_progress}
    unstarted = [j for j in journeys if j.get("id") not in started_ids]
    target = unstarted[0] if unstarted else journeys[0]
    
    return {
        "journey": target,
        "reason": f"Matches spiritual growth path in {target.get('category', 'Faith')}",
        "personalized": True
    }

File: microservices/domain-2/service-b/src/agents/journeys.ts
import fs from 'node:fs';
import path from 'node:path';
import { db } from '../db';
import { Journey, JourneyDay, UserJourneyProgress, Mood } from '../types';
import { moodHistory } from './journal';
import { JourneyRepository } from '../repositories';

let catalogCache: { journeys: Journey[]; days: JourneyDay[] } | null = null;
function getCatalog(): { journeys: Journey[]; days: JourneyDay[] } {
  if (!catalogCache) {
    catalogCache = JSON.parse(
      fs.readFileSync(path.join(__dirname, '..', 'data', 'journeys.json'), 'utf-8')
    );
  }
  return catalogCache!;
}

export async function listJourneys(): Promise<Journey[]> {
  const supabaseJourneys = await JourneyRepository.listAll();
  if (supabaseJourneys && supabaseJourneys.length > 0) {
    return supabaseJourneys;
  }
  return getCatalog().journeys;
}

export async function getJourney(journeyId: string): Promise<Journey | null> {
  const journeys = await listJourneys();
  return journeys.find(j => j.id === journeyId) || null;
}

export async function getJourneyDay(journeyId: string, dayNumber: number): Promise<JourneyDay | null> {
  const supabaseDay = await JourneyRepository.getDay(journeyId, dayNumber);
  if (supabaseDay) return supabaseDay;
  return getCatalog().days.find(d => d.journeyId === journeyId && d.dayNumber === dayNumber) || null;
}

export async function listUserJourneys(userId: string): Promise<UserJourneyProgress[]> {
  return JourneyRepository.getUserProgress(userId);
}

export async function startJourney(userId: string, journeyId: string): Promise<UserJourneyProgress> {
  const journey = await getJourney(journeyId);
  if (!journey) throw new Error('Journey not found');

  const existingList = await listUserJourneys(userId);
  const existing = existingList.find(p => p.journeyId === journeyId);
  if (existing) return existing;

  const progress: UserJourneyProgress = {
    journeyId,
    currentDay: 1,
    completedDays: [],
    startedAt: new Date().toISOString(),
  };

  await JourneyRepository.syncUserProgress({
    userId,
    journeyId,
    currentDay: progress.currentDay,
    completedDays: progress.completedDays,
  });

  return progress;
}

export async function getActiveJourney(userId: string): Promise<{ journey: Journey; progress: UserJourneyProgress; day: JourneyDay } | null> {
  const list = (await listUserJourneys(userId)).filter(p => !p.completedAt);
  if (list.length === 0) return null;
  const progress = list[list.length - 1];
  const journey = await getJourney(progress.journeyId);
  const day = journey ? await getJourneyDay(progress.journeyId, progress.currentDay) : null;
  if (!journey || !day) return null;
  return { journey, progress, day };
}

export async function completeDay(userId: string, journeyId: string): Promise<UserJourneyProgress> {
  const list = await listUserJourneys(userId);
  const progress = list.find(p => p.journeyId === journeyId);
  if (!progress) throw new Error('Journey not started for this user');

  const journey = await getJourney(journeyId);
  if (!journey) throw new Error('Journey not found');

  if (!progress.completedDays.includes(progress.currentDay)) {
    progress.completedDays.push(progress.currentDay);
  }

  if (progress.currentDay >= journey.totalDays) {
    progress.completedAt = new Date().toISOString();
  } else {
    progress.currentDay += 1;
  }

  await JourneyRepository.syncUserProgress({
    userId,
    journeyId,
    currentDay: progress.currentDay,
    completedDays: progress.completedDays,
    completedAt: progress.completedAt,
  });

  return progress;
}

export interface JourneyRecommendation {
  journey: Journey;
  reason: string;
  personalized: boolean;
}

const MOOD_STATE: Record<Mood, string> = {
  grateful: 'feeling grateful',
  peaceful: 'feeling peaceful',
  seeking: 'in a season of seeking',
  doubting: 'wrestling with doubt',
  distant: 'feeling distant from God',
  convicted: 'feeling convicted',
};

export async function getRecommendedJourney(userId: string): Promise<JourneyRecommendation | null> {
  const catalog = await listJourneys();
  const userProgress = await listUserJourneys(userId);
  const started = new Set(userProgress.map(p => p.journeyId));
  const candidates = catalog.filter(j => !started.has(j.id));
  if (candidates.length === 0) return null;

  const history = await moodHistory(userId, 7);
  const counts: Partial<Record<Mood, number>> = {};
  for (const entry of history) {
    if (entry.mood) counts[entry.mood as Mood] = (counts[entry.mood as Mood] || 0) + 1;
  }

  let best: Journey | null = null;
  let bestScore = 0;
  let bestMood: Mood | null = null;

  for (const journey of candidates) {
    for (const mood of journey.recommendedMoods) {
      const score = counts[mood as Mood] || 0;
      if (score > bestScore) {
        bestScore = score;
        best = journey;
        bestMood = mood as Mood;
      }
    }
  }

  if (best && bestScore > 0) {
    return {
      journey: best,
      reason: `Because you've been ${MOOD_STATE[bestMood!]} this week`,
      personalized: true,
    };
  }

  return {
    journey: candidates[0],
    reason: history.every(h => !h.mood) ? 'A good place to start' : 'Something new to try',
    personalized: false,
  };
}
